### ISSUE TITLE
Multiple issues in the new GraphQL transformer migration process.  

### ISSUE THREAD

**Issue by kylekirkby** [aws-amplify/amplify-cli#9200]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

14.17.5

### Amplify CLI Version

7.6.2

### What operating system are you using?

Windows 10 / Ubuntu 20.04 WSL

### Amplify Categories

api

### Amplify Commands

push

### Describe the bug

So I'm working on migrating to the new GraphQL transformer and I've been hitting a series of issues. I've run `amplify migrate api` and made the recommended changes after running `amplify api gql-compile` until the schema compiles. 

When running `amplify push` I was first hit with a new issue surrounding the maximum number of IAM policies per role. This was not the case before the migration to the new transformer version. I requested an increase for IAM Managed Policies per Role from 10 to 20 and it was approved within a couple of minutes. I ran `amplify push` again to see another error...

The next error I hit was surrounding `Only one resolver is allowed per field`. Reading through GitHub issues, someone mentioned trying to run `amplify push` again. That resolved that issue, onto the next...

`Data source with name NONE_DS` is now preventing `amplify push` from succeeding. I'm going to try and delete the `NONE_DS` data source in the AppSync console and see if that unblocks me. 

In short, there are multiple issues with the migration process that need some attention. None of these issues happened before the migration. `amplify api gql-compile` succeeds so if there transpires to be issues with the schema, then the compilation logic should probably be improved to prevent these issues from occuring.

🤞🏼 that deleting the NONE_DS data source will unblock me.

### Expected behavior

As above.

### Reproduction steps

1. As above.

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by kylekirkby** [comment]
Update:

I cannot delete the `NONE_DS` data source as `Error: Data source is still in use by functions`..

**Comment by kylekirkby** [comment]
So I've managed to get past this issue by going to the 9 individual AppSync functions that were still using the `NONE_DS` data source and changing the data source to a different NONE type data source. 

It has failed again with a different error... This time nothing verbose is displayed in the Cloudformation logs so waiting for Cloudformation to finish now to see . 🤦🏼‍♂️ It's been stuck on `functionAdminQueries` for some time now....

**Comment by kylekirkby** [comment]
So it took a while but finally completed the rollback. I haven't got a specific error to go by now. All I've got is:
```
The following resource(s) failed to update: [ResourceTheme, EmailTranslation, ResourceTag,.......]
```

**Comment by kylekirkby** [comment]
@Straubulous @marcvberg @sundersc - any idea on what's causing these issues? We can't make use of the new features/improvements in v2 until we can get these resolved. Thanks in advance!

**Comment by marcvberg** [comment]
Hi @kylekirkby,

I'm getting into this today, to replicate this could you describe which directives/features you were using with your API with the V1 transformer? I'm going to try to put together a V1 schema that replicates your issues above

**Comment by kylekirkby** [comment]
We've used field-level auth rules, field-level function resolvers, custom mutations, many-to-many connection tables, multiple auth rules, custom `@key`'s etc. Let me know if you need more details.

**Comment by marcvberg** [comment]
I'll look into this - in the meantime, have you reviewed the migration guide? The overall announcement is here: https://docs.amplify.aws/cli/migration/transformer-migration/

and referenced within this page are the things that require manual migration: https://docs.amplify.aws/cli/migration/transformer-migration/#what-are-the-breaking-changes-and-how-do-i-migrate

Custom mutations come into the custom resolvers category, so if you're relying on `amplify migrate api` to complete the migration and have your resources ready it won't be able to get everything

**Comment by kylekirkby** [comment]
@marcvberg - thanks... but I've already read the migration guide and migrated my `schema.graphql` file accordingly prior to creating this ticket.

**Comment by kylekirkby** [comment]
I've recently checked out my main environment and now upon trying to checkout my develop env I'm getting an error. This hasn't happened on the previous CLI version.

```
amplify env checkout develop
⠼ Initializing your environment: develop
Could not initialize 'develop': undefined
```
I was about to send over my amplify folder.. any ideas?

**Comment by kylekirkby** [comment]
It looks like my environment is changing since when I run `amplify status` it shows the environment I've just checked out. However, all resources are showing "Create" instead of "No change" etc.

**Comment by kylekirkby** [comment]
@marcvberg - I'm now in a fully blocked state. 
```
An error occurred during the push operation: null
```
I cannot push any changes or checkout different environments.

**Comment by marcvberg** [comment]
There's currently an AWS outage causing many amplify commands (among other things) to fail

**Comment by hackrx** [comment]
Hello @marcvberg, I added a custom query for a type, and updated `customResources.json` file, push is not working. I am continuously getting `only one resolver is allowed per field` ( still after 3 retries, as suggested in many comments), so I decided to migrate to v2.
I renamed my each resolver which was under `/api/<name/resolvers/` from `.req|res.vtl` to `.postAuth.1.req|res.vtl`, this is  what I need to do right? Do I need to do the same with pipeline resolvers?  does pipeline resolver also requires slot value`{init, preAuth, auth, postAuth, preDataLoad}`?
one more thing I want to clarify, do I need to also update `CustomResources.json` under `/api/stack/` for both cases?

**Comment by kylekirkby** [comment]
Hi @hackrx ,

Could you please create a separate issue for your issue?

Thanks,

Kyle

**Comment by sundersc** [comment]
Hi @kylekirkby - Can you try to pull the project to a new folder and email us the `amplify` folder to  amplify-cli@amazon.com?

**Comment by prafullatandel** [comment]
Hi @marcvberg 
We are having same issue, after using "amplify Migrate api"  the schema.graphql gets migrated to V2  But upon using amplify push we to get the same error. **Data source with name NONE_DS already exists**

"CREATE_FAILED      GraphQLAPINONEDS95A13CF0            AWS::AppSync::DataSource    Tue Dec 14 2021 08:51:02 GMT+0000 (Coordinated Universal Time) **Data source with name NONE_DS already exists** (Service: AWSAppSync;
Status Code: 400; Error Code: BadRequestException; Request ID: 8228de85-74c3-49bb-bf88-1be495278f96; Proxy: null)"

We do not have any custom resolvers.

Do we have any solution for this issue.

Regards,
Prafulla

**Comment by sundersc** [comment]
Hi @prafullatandel - Can you post your schema so that we can try to reproduce it on our end? Also, let us know the cli version.

**Comment by frankclaassen** [comment]
I get the same issue which seems to be due to having DataStore enabled

**Comment by skorobahatko** [comment]
Have the same issue after changing schema to v2, amplify push isn't working

**Comment by sundersc** [comment]
@frankclaassen @skorobahatko - Could you share the graphql schema?

**Comment by frankclaassen** [comment]
Doubt it will help you if I send my amplify folder as I have managed to get past this issue mostly, just stuck with an error now which I will post further down. To overcome most issues, I had to update auth first and push, then go onto the appsync console to manually update all resources pointing to NONE_DS, delete NONE_DS datasource, run api migrate and push.

Now I am getting the below error with the migrated schema:

✖ An error occurred when pushing the resources to the cloud
🛑 ["Index: 0 State: {\"deploy\":\"waitingForDeployment\"} Message: Resource is not in the state stackUpdateComplete"]
DeploymentError: ["Index: 0 State: {\"deploy\":\"waitingForDeployment\"} Message: Resource is not in the state stackUpdateComplete"]
    at /usr/local/lib/node_modules/@aws-amplify/cli/node_modules/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:159:40
    at Interpreter.update (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/interpreter.js:265:9)
    at /usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/interpreter.js:110:15
    at Scheduler.process (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/scheduler.js:69:7)
    at Scheduler.flushEvents (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/scheduler.js:60:12)
    at Scheduler.schedule (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/scheduler.js:49:10)
    at Interpreter.send (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/interpreter.js:104:23)
    at actor.id (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/xstate/lib/interpreter.js:1010:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

**Comment by frankclaassen** [comment]
The above is on the latest version of the cli

**Comment by sundersc** [comment]
@frankclaassen The error 'Resource is not in the state stackUpdateComplete' usually pops when there is an ongoing update on your stack. Can you check the status of your application stack in cloudformation (aws console) and make sure it is 'UPDATE_COMPLETE'? Then try `amplify pull`, make the necessary updates and do `amplify push`.

**Comment by kylekirkby** [comment]
Hi all,

So we are back to trying to migrate to the new version of the GraphQL transformer. I'm still hitting the issue regarding:

```
Data source with name NONE_DS already exists
```

I'm on version 7.6.12 of the CLI.

Any help would be appreciated since we need to migrate now to use new authorization features that exist in the latest version.

Thanks in advance,

Kyle

**Comment by kylekirkby** [comment]
Update:

I thought I was on to a workaround using the CDK overrides for the AppSync API to change the name of the NONE_DS DataSource to e.g `NEW_DONE_DS`.

However, according to the Amplify docs, this **should** be possible by doing something like:

```
import { AmplifyApiGraphQlResourceStackTemplate } from '@aws-amplify/cli-extensibility-helper';

export function override(resources: AmplifyApiGraphQlResourceStackTemplate) {
	resources.api.GraphQLAPINONEDS.addPropertyOverride('Properties.Name', 'NEW_DONE_DS');
}
```

But even after deleting my build/ folder for my API and generating an entirely new build with `amplify api gql-compile` my NONE DataSource is named: `GraphQLAPINONEDS95A13CF3`. Thus resulting in a typescript error and yet another failed stack deployment.

**Comment by sundersc** [comment]
@kylekirkby - Could you include the graphql schema in the issue template?

**Comment by kylekirkby** [comment]
@sundersc - Just to confirm, I emailed this over last night.

**Comment by sundersc** [comment]
Discussed offline. Here are the steps followed to resolve the issue.

- Update the schema to V1 schema and do amplify push, wait for it to complete.
- Go to AppSync console and open the api. For all the functions using NONE_DS datasource, manually change it to NONE.
- Delete the NONE_DS datasource.
- Update the schema to v2 and push again.

Closing this issue now. Feel free to reopen if you encounter the same issue again.

**Comment by skybach** [comment]
> Discussed offline. Here are the steps followed to resolve the issue.
> 
> * Update the schema to V1 schema and do amplify push, wait for it to complete.
> * Go to AppSync console and open the api. For all the functions using NONE_DS datasource, manually change it to NONE.
> * Delete the NONE_DS datasource.
> * Update the schema to v2 and push again.
> 
> Closing this issue now. Feel free to reopen if you encounter the same issue again.

is there an update to the AppSync AWS Console? there is no NONE option.
